nativelib/src/main/scala/scala/scalanative/unsafe/CFuncPtr.scala

```diff
@@ -11,78 +11,213 @@
 package scala.scalanative
 package unsafe
 
-sealed trait CFuncPtr
+import scala.language.implicitConversions
+import scala.scalanative.runtime.Boxes._
+import scala.scalanative.runtime.{RawPtr, intrinsic}
 
-// ###sourceLocation(file: "nativelib/src/main/scala/scala/scalanative/unsafe/CFuncPtr.scala.gyb", line: 17)
+sealed trait CFuncPtr {
+  def rawptr: RawPtr
+}
 
-trait CFuncPtr0[R] extends CFuncPtr {
-  def apply(): R
+final class CFuncRawPtr(val rawptr: RawPtr) extends CFuncPtr {
+  def as[F <: CFuncPtr](implicit tag: Tag.CFuncPtrTag[F]): F =
+    tag.load(boxToPtr[F](rawptr), loadPtr = false)
 }
 
-// ###sourceLocation(file: "nativelib/src/main/scala/scala/scalanative/unsafe/CFuncPtr.scala.gyb", line: 17)
+// ###sourceLocation(file: "nativelib/src/main/scala/scala/scalanative/unsafe/CFuncPtr.scala.gyb", line: 36)
 
-trait CFuncPtr1[T1, R] extends CFuncPtr {
-  def apply(arg1: T1): R
+final class CFuncPtr0[R](val rawptr: RawPtr) extends CFuncPtr {
+  def apply()(implicit evRet: Tag[R]): R = intrinsic
 }
 
-// ###sourceLocation(file: "nativelib/src/main/scala/scala/scalanative/unsafe/CFuncPtr.scala.gyb", line: 17)
+object CFuncPtr0 {
+  implicit def fromScalaFunction0[R](fn: Function0[R]): CFuncPtr0[R] = intrinsic
+
+  def fromRawPtr[R](ptr: RawPtr): CFuncPtr0[R] = {
+    new CFuncPtr0[R](ptr)
+  }
+}
+// ###sourceLocation(file: "nativelib/src/main/scala/scala/scalanative/unsafe/CFuncPtr.scala.gyb", line: 36)
 
-trait CFuncPtr2[T1, T2, R] extends CFuncPtr {
-  def apply(arg1: T1, arg2: T2): R
+final class CFuncPtr1[T1, R](val rawptr: RawPtr) extends CFuncPtr {
+  def apply(arg1: T1)(implicit ev1: Tag[T1], evRet: Tag[R]): R = intrinsic
 }
 
-// ###sourceLocation(file: "nativelib/src/main/scala/scala/scalanative/unsafe/CFuncPtr.scala.gyb", line: 17)
+object CFuncPtr1 {
+  implicit def fromScalaFunction1[T1, R](
+      fn: Function1[T1, R]): CFuncPtr1[T1, R] = intrinsic
+
+  def fromRawPtr[T1, R](ptr: RawPtr): CFuncPtr1[T1, R] = {
+    new CFuncPtr1[T1, R](ptr)
+  }
+}
+// ###sourceLocation(file: "nativelib/src/main/scala/scala/scalanative/unsafe/CFuncPtr.scala.gyb", line: 36)
 
-trait CFuncPtr3[T1, T2, T3, R] extends CFuncPtr {
-  def apply(arg1: T1, arg2: T2, arg3: T3): R
+final class CFuncPtr2[T1, T2, R](val rawptr: RawPtr) extends CFuncPtr {
+  def apply(arg1: T1,
+            arg2: T2)(implicit ev1: Tag[T1], ev2: Tag[T2], evRet: Tag[R]): R =
+    intrinsic
 }
 
-// ###sourceLocation(file: "nativelib/src/main/scala/scala/scalanative/unsafe/CFuncPtr.scala.gyb", line: 17)
+object CFuncPtr2 {
+  implicit def fromScalaFunction2[T1, T2, R](
```

Comment:
Can we change the implicit convertions methods naming to `fromScalaFunction`? It should be cleaner and compiler still should distinguish different methods since they have different argument lists